### ISSUE TITLE
GCE provider issues

### ISSUE THREAD

**Issue by afgane** [CloudVE/cloudbridge#32]
In testing PR #31, I came across a few issues with the GCE provider so will summarize those here to keep track of them (@chiniforooshan, @baizhang, @mbookman):

- [x]  `KeyError: 'methods'` when creating a provider
Cannot create a GCE provider because getting the following exception:
```
/Users/eafgan/projects/pprojects/cloudbridge-branches/gce/cloudbridge/cloud/providers/gce/provider.pyc in __init__(self, connection)
     83         # We will not mutate self._desc; it's OK to use items() in Python 2.x.
     84         for resource, resource_desc in desc['resources'].items():
---> 85             methods = resource_desc['methods']
     86             if 'get' not in methods:
     87                 continue

KeyError: 'methods'
```
This happens on both py27 and py36 with a provider config like this:
```
gce_config = {'gce_project_name': 'project name',
                       'gce_service_creds_file': 'service acct file.json'}
```
Simply changing line 85 in gce `provider.py` to `methods = resource_desc.get('methods', "")` resolves this but I'm not sure if that's the intended path.

- [x] `TypeError: Missing required parameter "zone"` when operating on some compute services
When trying to run the following commands (possibly others) `provider.compute.instances.list()` or `provider.compute.instance_types.list()`, getting the following traceback:
```
/Users/eafgan/.pyenv/versions/cb-py2/lib/python2.7/site-packages/googleapiclient/discovery.pyc in method(self, **kwargs)
    725     for name in parameters.required_params:
    726       if name not in kwargs:
--> 727         raise TypeError('Missing required parameter "%s"' % name)
    728
    729     for name, regex in six.iteritems(parameters.pattern_params):

TypeError: Missing required parameter "zone"
```
Explicitly defining `gce_default_zone` in the provider config solves this but cannot we have a default default?

- [x] `instances.list` method does not work when no instances are present
Just running `provide.compute.instances.list()` comes back with this when there are no live instances currently available:
```
/Users/eafgan/projects/pprojects/cloudbridge-branches/gce/cloudbridge/cloud/providers/gce/services.pyc in list(self, limit, marker)
    524             pageToken=marker).execute()
    525         instances = [GCEInstance(self.provider, inst)
--> 526                      for inst in response['items']]
    527         return ServerPagedResultList(len(instances) > max_result,
    528                                      response.get('nextPageToken'),

KeyError: 'items'
```

- [x] `instance.create` method does not return an `Instance` object of the instance that was just created.

- [ ] `instance.refresh()` or `provider.compute.instances.get(inst_id)` causes an exception:
```
/Users/eafgan/.pyenv/versions/cb-py2/lib/python2.7/site-packages/oauth2client/client.pyc in _do_refresh_request(self, http)
    810             except (TypeError, ValueError):
    811                 pass
--> 812             raise HttpAccessTokenRefreshError(error_msg, status=resp.status)
    813
    814     def _revoke(self, http):

HttpAccessTokenRefreshError: invalid_scope: Empty or missing scope not allowed.
```

-[ ] On `instance.create(name='ea-pr', image=img, instance_type=it, subnet=sn)` call, getting 
```
/Users/eafgan/projects/pprojects/cloudbridge-branches/gce/cloudbridge/cloud/providers/gce/services.py in create(self, name, image, instance_type, subnet, zone, key_pair, security_groups, user_data, launch_config, **kwargs)
    495                          .insert(project=self.provider.project_name,
    496                                  zone=self.provider.default_zone,
--> 497                                  body=config)
    498                          .execute())
    499         if 'zone' not in operation:

HttpError: <HttpError 400 when requesting https://www.googleapis.com/compute/v1/projects/isb-cgc-03-0002/zones/us-east1-b/instances?alt=json returned "Invalid value for field 'resource.networkInterfaces[0]': ''. Subnetwork should be specified for custom subnetmode network">
```
This TODO is probably to blame: https://github.com/gvlproject/cloudbridge/blob/gce/cloudbridge/cloud/providers/gce/services.py#L476

**Comment by nuwang** [comment]
I think this would also be a great opportunity to watch out for ways to harden the test suite. I think some of these may not be getting caught in tests (e.g. the list method() not working when no instances exist, although that's a bad example, because there's no reliable way to test for that in a non-empty project.)

**Comment by mbookman** [comment]
`KeyError: 'methods'` should be fixed. The code now indicates:

  methods = resource_desc.get('methods', {})

**Comment by mbookman** [comment]
`instance.create method does not return an Instance object of the instance that was just created.`: is fixed.

**Comment by afgane** [comment]
Thanks for following up here. I've updated the issue list. While all but one listed issue were resolved,  I did add one more now that I encountered when trying to launch an instance. I see there's a TODO lingering (https://github.com/gvlproject/cloudbridge/blob/gce/cloudbridge/cloud/providers/gce/services.py#L476) so I'm just going to guess that's causing this but figured it was good to include here to keep track of it easier.
